### PR TITLE
Issue #95 refactor path planner tests : Parameterized path planner tests

### PR DESCRIPTION
**Pull request by madratman**
This PR addresses the first point raised in the issue. Please review.

### CODE REVIEW

**Comment by madratman** [comment]
Okay, my fork's hydro-devel was lagging behind ros-industrial's hydro-devel it seems, as I am not able to catkin_make (PR https://github.com/ros-industrial-consortium/descartes/pull/101)

**Comment by madratman** [comment]
There are some problems here. Looking into them
`[ 92%] Built target descartes_moveit
[100%] Building CXX object descartes/descartes_planner/CMakeFiles/descartes_planner.dir/test/TestPoint.cpp.o
/home/ratnesh/catkin_ws_exp/src/descartes/descartes_planner/test/TestPoint.cpp:37:44: error: ‘RobotModel’ does not name a type
/home/ratnesh/catkin_ws_exp/src/descartes/descartes_planner/test/TestPoint.cpp:37:56: error: ISO C++ forbids declaration of ‘model’ with no type [-fpermissive]
/home/ratnesh/catkin_ws_exp/src/descartes/descartes_planner/test/TestPoint.cpp:45:36: error: ‘RobotModel’ does not name a type
/home/ratnesh/catkin_ws_exp/src/descartes/descartes_planner/test/TestPoint.cpp:45:48: error: ISO C++ forbids declaration of ‘model’ with no type [-fpermissive]
make[2]: *** [descartes/descartes_planner/CMakeFiles/descartes_planner.dir/test/TestPoint.cpp.o] Error 1
make[1]: *** [descartes/descartes_planner/CMakeFiles/descartes_planner.dir/all] Error 2
make: *** [all] Error 2
Invoking "make" failed`

And I forgot to create a test trajectory!

**Comment by shaun-edwards** [comment]
@madratman, just ping the PR, when it is ready for review.

**Comment by madratman** [comment]
@shaun-edwards  Upon running `$ catkin_make run_tests_descartes_planner_gtest`, targets are getting built, though while linking:

`Linking CXX executable /home/ratnesh/catkin_ws/devel/lib/descartes_planner/descartes_planner_utest
CMakeFiles/descartes_planner_utest.dir/test/planner_test.cpp.o: In function`ThreeDOFRobot::ThreeDOFRobot()':
planner_test.cpp:(.text._ZN13ThreeDOFRobotC2Ev[_ZN13ThreeDOFRobotC5Ev]+0x2c): undefined reference to `vtable for ThreeDOFRobot'
CMakeFiles/descartes_planner_utest.dir/test/planner_test.cpp.o: In function`TestPoint::TestPoint(std::vector<double, std::allocator<double> > const&)':
planner_test.cpp:(.text._ZN9TestPointC2ERKSt6vectorIdSaIdEE[_ZN9TestPointC5ERKSt6vectorIdSaIdEE]+0x24): undefined reference to `vtable for TestPoint'
collect2: ld returned 1 exit status
make[3]: *** [/home/ratnesh/catkin_ws/devel/lib/descartes_planner/descartes_planner_utest] Error 1
make[2]: *** [descartes/descartes_planner/CMakeFiles/descartes_planner_utest.dir/all] Error 2
make[1]: *** [descartes/descartes_planner/CMakeFiles/run_tests_descartes_planner_gtest.dir/rule] Error 2
make: *** [run_tests_descartes_planner_gtest] Error 2
Invoking "make" failed
`

http://stackoverflow.com/questions/3065154/undefined-reference-to-vtable says that this error might occur if any virtual functon that isn't pure isn't defined, including destructor. But, I can't see what is missing here.

**Comment by Jmeyer1292** [comment]
Did you add all your source files for the tests to the appropriate spot at the bottom of the CMakeLists.txt for the planner?

**Comment by madratman** [comment]
Okay I added 
`test/TestPoint.cpp 
test/ThreeDOFRobot.cpp`
at https://github.com/madratman/descartes/blob/issue_%2395_refactor_path_planner_tests/descartes_planner/CMakeLists.txt/#L81.

Now TestPoint.cpp.o and ThreeDOFRobot.cpp.o were built, but I am getting the exact same error again. 

Also, I should put TestPoint and ThreeDOFRobot under the descartes_planner_test namespace?

**Comment by madratman** [comment]
After doing this("this" here referes to putting TestPoint and ThreeDOFRobot under the descartes_planner_test namespace), now I'm getting the vtable error only for TestPoint (but why?):
`Linking CXX executable /home/ratnesh/catkin_ws/devel/lib/descartes_planner/descartes_planner_utest
CMakeFiles/descartes_planner_utest.dir/test/planner_test.cpp.o: In function`descartes_planner_test::TestPoint::TestPoint(std::vector<double, std::allocator<double> > const&)':
planner_test.cpp:(.text._ZN22descartes_planner_test9TestPointC2ERKSt6vectorIdSaIdEE[_ZN22descartes_planner_test9TestPointC5ERKSt6vectorIdSaIdEE]+0x24): undefined reference to `vtable for descartes_planner_test::TestPoint'
collect2: ld returned 1 exit status
make[3]: *** [/home/ratnesh/catkin_ws/devel/lib/descartes_planner/descartes_planner_utest] Error 1
make[2]: *** [descartes/descartes_planner/CMakeFiles/descartes_planner_utest.dir/all] Error 2
make[1]: *** [descartes/descartes_planner/CMakeFiles/run_tests_descartes_planner_gtest.dir/rule] Error 2
make: *** [run_tests_descartes_planner_gtest] Error 2
Invoking "make" failed`

File: descartes_planner/CMakeLists.txt [hunk 1]
```diff
@@ -47,6 +47,8 @@ add_library(descartes_planner
             src/sparse_planner.cpp
             src/dense_planner.cpp
             src/plugin_init.cpp
+            test/ThreeDOFRobot.cpp
```
Comment (madratman) [inline descartes_planner/CMakeLists.txt]: We don't need L50-L51, right?

**Comment by madratman** [comment]
From http://stackoverflow.com/questions/9406580/c-undefined-reference-to-vtable-and-inheritance, we do have the CartTrajectoryPt destructor defined [here](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_trajectory/include/descartes_trajectory/cart_trajectory_pt.h/#L246).
Do we need it [here](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_trajectory/src/cart_trajectory_pt.cpp) as well?

**Comment by shaun-edwards** [comment]
@madratman, I'll take a look this morning.

**Comment by shaun-edwards** [comment]
Not sure this is the problem, but you have TestPoint defined twice.  In a source file descartes/descartes_planner/test/TestPoint.cpp and descartes/descartes_planner/test/descartes_planner_test/TestPoint.h.  Normally, you would not see the class declaration in a source file, like this in TestPoint.cpp

```
class TestPoint: public descartes_trajectory::CartTrajectoryPt
  {
  public:
...
```

**Comment by madratman** [comment]
Yes, thanks,  this was the problem. Tests are failing though right now. I am unable
to give much time as end terms are going on. Will get back to it as soon as
they are over. Sorry.

On Mon, Apr 20, 2015 at 8:36 PM, Shaun Edwards notifications@github.com
 wrote:

> Not sure this is the problem, but you have TestPoint defined twice. In a
> source file descartes/descartes_planner/test/TestPoint.cpp and
> descartes/descartes_planner/test/descartes_planner_test/TestPoint.h.
> Normally, you would not see the class declaration in a source file, like
> this in TestPoint.cpp
> 
> class TestPoint: public descartes_trajectory::CartTrajectoryPt
>   {
>   public:
> ...
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-industrial-consortium/descartes/pull/111#issuecomment-94477728
> .
> 
> Ratnesh Madaan
> Senior Year
> B.Tech. Mechanical Engineering
> Indian Institute of Technology, Roorkee

**Comment by shaun-edwards** [comment]
@madratman  Is this PR still active or should it be closed?

**Comment by madratman** [comment]
Sorry, forgot about this. I can look it into once I am able to get the Hamilton PR#1 merged.

**Comment by madratman** [comment]
@shaun-edwards Ready for review. 
Just one point - the indigo-devel branch's dense_planner_test checks for 100 points and for sparse the CartID runs till 1000 but jumps 10. 
With the param-tests, I see 1000 points each.

**Comment by madratman** [comment]
All tests passing now. But the no of test trajectory points (`NUM_DENSE_POINTS`) is same for both sparse and dense planner tests. 
Pass planner type as an argument -> value parameterized tests
How to combine type-parameterized and value-parameterized tests?
Doesn't seem feasible([combining Value-Parameterized test and Type-Parametered](https://groups.google.com/forum/#!topic/googletestframework/3d8H5J5Isnw))

Or simply check planner type ->how?

**Comment by shaun-edwards** [comment]
The planners should return the same number of points as the cartesian path that is passed to the planner.  Dense/sparse refers to how the planner solves.  In the case of sparse, it may sample the path, but it returns a solution with the full density.

I don't understand your other question.  Why do you feel you need a value-parameterized test?

**Comment by madratman** [comment]
Yes, I am aware of the difference. Just in the current implementation, [here](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_planner/test/dense_planner.cpp#L31) and [here](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_planner/test/sparse_planner.cpp#l31), the number of traj points were different, which seemed a typo. But anyway, you can just specify different numbers if needed. Or this is useless and I should revert the commits?

edit: It seems a bit useless now as the point of using type-parameterized in the first place was avoiding duplicate code. And this could have been achieved by running partial tests as well. 

edit: 

```
[ INFO] [1435722353.314401770]: CartID: ID8200 JointPoses count: 1
[ INFO] [1435722353.314586161]: CartID: ID8201 JointPoses count: 1
[ INFO] [1435722353.314708494]: CartID: ID8202 JointPoses count: 1
[ INFO] [1435722353.315509891]: Found 1000 joint solutions available
[ INFO] [1435722353.318397189]: Not calculating edge weights to nil ID
[ INFO] [1435722353.328533519]: With 1000 - new best score: 6.930746
[ INFO] [1435722353.334585739]: Tearing down DescartesPlannerTest fixture(base) (parameterized)
...
```

8202 is the upper limit it seems?

PS. Out of context, but the link to indigo-training points to the wiki [here](http://rosindustrial.org/news/2015/6/26/descartes-path-planner-alpha-release)

**Comment by madratman** [comment]
Moving on, to the other [three points](https://github.com/ros-industrial-consortium/descartes/issues/95) in the issue.

- jointVelocityLimit can be accessed from [here](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_moveit/src/moveit_state_adapter.cpp#L32).

**Comment by shaun-edwards** [comment]
Closing due to a lack of activity.

**Pull request closed without merging**